examples/tracing-http/src/server.rs

```diff
@@ -0,0 +1,107 @@
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use opentelemetry::{
+    global,
+    propagation::TextMapPropagator,
+    trace::{FutureExt, Span, SpanKind, TraceContextExt, Tracer},
+    Context,
+};
+use opentelemetry_contrib::trace::propagator::trace_context_response::TraceContextResponsePropagator;
+use opentelemetry_http::{HeaderExtractor, HeaderInjector};
+use opentelemetry_sdk::{propagation::TraceContextPropagator, trace::TracerProvider};
+use opentelemetry_stdout::SpanExporter;
+use std::{convert::Infallible, net::SocketAddr};
+
+// Utility function to extract the context from the incoming request headers
+fn extract_context_from_request(req: &Request<Body>) -> Context {
+    global::get_text_map_propagator(|propagator| {
+        propagator.extract(&HeaderExtractor(req.headers()))
+    })
+}
+
+// Separate async function for the handle endpoint
+async fn handle_health_check(_req: Request<Body>) -> Result<Response<Body>, Infallible> {
+    let tracer = global::tracer("example/server");
+    let mut span = tracer
+        .span_builder("health_check")
+        .with_kind(SpanKind::Server)
+        .start(&tracer);
+    span.add_event("Health check accessed", vec![]);
+    let cx = Context::default().with_span(span);
+    let mut res = Response::new(Body::from("Server is up and running!"));
+    let response_propagator: &dyn TextMapPropagator = &TraceContextResponsePropagator::new();
+    response_propagator.inject_context(&cx, &mut HeaderInjector(res.headers_mut()));
+    Ok(res)
+}
+
+// Separate async function for the echo endpoint
+async fn handle_echo(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+    let tracer = global::tracer("example/server");
+    let mut span = tracer
+        .span_builder("echo")
+        .with_kind(SpanKind::Server)
+        .start(&tracer);
+    span.add_event("Echoing back the request", vec![]);
+    let cx = Context::default().with_span(span);
+    let mut res = Response::new(req.into_body());
+    let response_propagator: &dyn TextMapPropagator = &TraceContextResponsePropagator::new();
+    response_propagator.inject_context(&cx, &mut HeaderInjector(res.headers_mut()));
+    Ok(res)
+}
+
+async fn router(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+    // Extract the context from the incoming request headers
+    let parent_cx = extract_context_from_request(&req);
+    let response = {
+        // Create a span parenting the remote client span.
+        let tracer = global::tracer("example/server");
+        let mut span = tracer
+            .span_builder("router")
+            .with_kind(SpanKind::Server)
+            .start_with_context(&tracer, &parent_cx);
+
+        span.add_event("dispatching request", vec![]);
+
+        let cx = Context::default().with_span(span);
+        match (req.method(), req.uri().path()) {
+            (&hyper::Method::GET, "/health") => handle_health_check(req).with_context(cx).await,
```

Comment:
what would have happened if we did not explicitly pass `cx` "with_context" ?

Reply:
The `handle_health_check` async function can run in separate thread, so `with_context` ensure the context (passed as argument) gets attached to the current_context during [polling ](https://github.com/open-telemetry/opentelemetry-rust/blob/47881b20a2b8e94d8e1cdbd4877852dd74cc07de/opentelemetry/src/trace/context.rs#L380)this future.

Reply:
Thanks for explaining. I see this is already documented as well : https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry/src/trace/tracer.rs#L73